docs/tutorials/fault_tolerance.md

```diff
@@ -0,0 +1,154 @@
+  purpose, such as a stateful redeploy/upgrade/rescaling operation. Savepoints are always complete,
+  and are optimized for operational flexibility.
+
+### How does State Snapshotting Work?
+
+Flink uses a variant of the [Chandy-Lamport
+algorithm](https://en.wikipedia.org/wiki/Chandy-Lamport_algorithm) known as _asynchronous barrier
+snapshotting_.
+
+When a task manager is instructed by the checkpoint coordinator (part of the job manager) to begin a
+checkpoint, it has all of the sources record their offsets and insert numbered _checkpoint barriers_
+into their streams. These barriers flow through the job graph, indicating the part of the stream
+before and after each checkpoint. 
+
+<img src="{{ site.baseurl }}/fig/stream_barriers.svg" alt="Checkpoint barriers are inserted into the streams" class="center" width="80%" />
+
+Checkpoint _n_ will contain the state of each operator that resulted from having consumed **every
+event before checkpoint barrier _n_, and none of the events after it**.
+
+As each operator in the job graph receives one of these barriers, it records its state. Operators
+with two input streams (such as a CoProcessFunction) perform _barrier alignment_ so that the
+snapshot will reflect the state resulting from consuming events from both input streams up to (but
+not past) both barriers.
+
+<img src="{{ site.baseurl }}/fig/stream_aligning.svg" alt="Barrier alignment" class="center" width="80%" />
+
+Flink's state backends use a copy-on-write mechanism to allow stream processing to continue
+unimpeded while older versions of the state are being asynchronously snapshotted. Only when the
+snapshots have been durably persisted will these older versions of the state be garbage collected.
+
+### Exactly Once Guarantees
+
+When things go wrong in a stream processing application, it's possible to have either lost, or
+duplicated results. With Flink, depending on the choices you make for your application and the
+cluster you run it on, any of these outcomes is possible:
+
+- Flink makes no effort to recover from failures (_at most once_)
+- Nothing is lost, but you may experience duplicated results (_at least once_)
+- Nothing is lost or duplicated (_exactly once_)
+
+Given that Flink recovers from faults by rewinding and replaying the source data streams, when the
+ideal situation is described as **exactly once** this does *not* mean that every event will be
+processed exactly once. Instead, it means that _every event will affect the state being managed by
+Flink exactly once_. 
+
+If you don't need exactly once semantics, you can gain some performance by disabling barrier
+alignment. This is done by configuring Flink to use `CheckpointingMode.AT_LEAST_ONCE`.
+
+### Exactly Once End-to-end
+
+To achieve exactly once end-to-end, so that every event from the sources affects the sinks exactly
+once, the following must be true:
+
+1. your sources must be replayable, and
+2. your sinks must be transactional (or idempotent)
```

Comment:
I think, this section may deserve a bit more content (not now, but definitely in the future) - can you create a ticket for that?

Reply:
https://issues.apache.org/jira/browse/FLINK-16216